### PR TITLE
Add an archive_enrollments service and worker

### PR DESCRIPTION
**Pull request by Isalafont**
- Add a service to archive enrollment with reminder events over 15 days
- Add spec
- Add worker + spec
- Add a cron job at 3am

### CODE REVIEW

**Comment by Isalafont** [comment]
[ETQ U, si j'ai reçu un email me prévenant de l'archivage prochaine d'une habilitation en statut "Brouillon", mon habilitation est effectivement archivée](https://trello.com/c/hX4UaySz/1741-etq-u-si-jai-re%C3%A7u-un-email-me-pr%C3%A9venant-de-larchivage-prochaine-dune-habilitation-en-statut-brouillon-mon-habilitation-est-effec)

File: backend/app/models/event.rb [hunk 1]
```diff
@@ -5,7 +5,7 @@ class Event < ActiveRecord::Base
   belongs_to :enrollment
 
   belongs_to :user, optional: true
-  validates :user, presence: true, if: proc { |event| event.name != "reminder" }
+  validates :user, presence: true, if: proc { |event| event.name != "reminder" && event.name != "archive" }
```
Comment (skelz0r) [inline backend/app/models/event.rb:8]:
```suggestion
  validates :user, presence: true, if: proc { |event| %w[reminder archive].exclude?(event.name) }
```

You should make this change in a single commit imo

File: backend/app/services/archive_draft_enrollments.rb [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+      .select { |event| event.name == "reminder" }
+      .pluck(:enrollment_id)
+
+    Enrollment.find(enrollment_ids)
+  end
+end
+
+# archived_enrollments.each do |enrollment|
+#   enrollment.update!(status: "archived")
+#   enrollment.events.create(name: "archive")
```
Comment (skelz0r) [inline backend/app/services/archive_draft_enrollments.rb:32]:
```suggestion
```

File: backend/app/services/archive_draft_enrollments.rb [hunk 3]
```diff
@@ -0,0 +1,32 @@
+# frozen_string_literal: true
+
+class ArchiveDraftEnrollments
```
Comment (skelz0r) [inline backend/app/services/archive_draft_enrollments.rb:3]: This service should be a query object imo.
A query object should return an activerecord relation, which can be manipulate with more query methods.

You can make a single SQL query here btw. If you want you can use https://www.db-fiddle.com/ to build you query.

Moreover, it's a bad name: this service does not archive draft enrollments, it sorts enrollments draft that can be archived

=> `DraftEnrollmentsArchiveable`

File: backend/app/services/archive_draft_enrollments.rb [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+  attr_reader :enrollment
```
Comment (skelz0r) [inline backend/app/services/archive_draft_enrollments.rb:4]: this line should be remove: there is no `enrollment` instance var here

File: backend/app/services/archive_draft_enrollments.rb [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+  end
+
+  def enrollments_to_archive
+    enrollments = select_enrollments_to_archive
+    enrollment_ids = enrollments.lazy.map { |enrollment| enrollment.events.last }.to_a
+      .select { |event| event.name == "reminder" }
+      .pluck(:enrollment_id)
+
+    Enrollment.find(enrollment_ids)
+  end
```
Comment (skelz0r) [inline backend/app/services/archive_draft_enrollments.rb:27]: I think you can reduce here with a single sql call

File: backend/app/services/archive_enrollments.rb [hunk 6]
```diff
@@ -0,0 +1,28 @@
+# frozen_string_literal: true
+
+class ArchiveEnrollments
```
Comment (skelz0r) [inline backend/app/services/archive_enrollments.rb:3]: Why did you split in 2 classes ? There is 2 options:

1. You need to make a disjonction: you can pass an argument (`statuses`) to the class, and pass it to your sql query
2. You don't need to make a disjonction: just use all statuses you want.

**Review by skelz0r — CHANGES_REQUESTED** [review]
You can refactor your 2 service object into a single query object like with the following interface:

```ruby
class EnrollmentsArchiveable
  def initialize(statuses=%w[draft changes_requested])
    @statuses = statuses
  end

  def perform
    # make the single sql query with having / subquery
  end
end
```

And in your worker:

```ruby
  def perform
    EnrollmentsArchiveable.new.perform.map(&:archive!)
  end
```

in the `Enrollment` model:

```ruby
  def archive!
      enrollment.update!(status: "archived")
      enrollment.events.create!(name: "archive")
  end
```

(this one should be tested too btw)

File: backend/app/services/archive_draft_enrollments.rb (hunk 3 above)
Comment (Isalafont) [reply backend/app/services/archive_draft_enrollments.rb:3]: This service file is supposed to be delete. 
I've made another service so we could archive enrollments according different status

I will check your link to improve the query methods

File: backend/app/services/archive_enrollments.rb (hunk 6 above)
Comment (Isalafont) [reply backend/app/services/archive_enrollments.rb:3]: Should I rename this class `EnrollmentsArchiveable` ?

Comment (skelz0r) [reply backend/app/services/archive_enrollments.rb:3]: yes

**Review by rdubigny — APPROVED** [review]

**Pull request merged**